bauble/src/value/display.rs

```diff
@@ -0,0 +1,593 @@
+                                        l.write(": ");
+                                        value.indented_display(l.reborrow());
+                                        l.write(",");
+                                    });
+                                }
+                            });
+                        }
+                        w.write("}");
+                    }
+                }
+            }
+            Value::Or(or) => match or.as_slice() {
+                [] => w.write("|"),
+                [variant] => {
+                    w.write_type_path();
+                    w.fmt(variant);
+                }
+                [rest @ .., last] => {
+                    for v in rest {
+                        w.write_type_path();
+                        w.fmt(v);
+                        w.write(" | ");
+                    }
+                    w.fmt(last);
+                }
+            },
+            Value::Primitive(primitive_value) => match primitive_value {
+                super::PrimitiveValue::Num(v) => w.fmt(v),
+                super::PrimitiveValue::Str(v) => w.debug_fmt(v),
+                super::PrimitiveValue::Bool(v) => w.fmt(v),
+                super::PrimitiveValue::Unit => w.write("()"),
+                // `Raw` is a bit annoying to handle here, since we don't know how it
+                // was originally expressed. So we need to check:
+                // 1. Can it be displayed as a literal raw, i.e `#some_raw`
+                // 2. If not, how many braces do we need, since raw values can contain
+                //    `}`, but it has to end with as many `}` as there was `{` at the
+                //    start.
+                super::PrimitiveValue::Raw(v) => {
+                    let mut num_braces = None;
+                    let mut row = None;
+
+                    let mut first = None;
+                    let mut last = None;
+                    for c in v.chars() {
+                        if first.is_none() {
+                            first = Some(c);
+                        }
+                        last = Some(c);
+                        if num_braces.is_none() && !allowed_in_raw_literal(c) {
+                            num_braces = Some(1);
+                        }
+
+                        if c == '}' {
+                            *row.get_or_insert(0) += 1;
+                        } else if let Some(r) = row {
+                            let current = num_braces.get_or_insert(r + 1);
+                            *current = (r + 1).max(*current);
+                            row = None;
+                        }
+                    }
+
+                    if let Some(n) = num_braces {
+                        w.write("#");
+                        for _ in 0..n {
+                            w.write("{");
+                        }
+                        if first == Some('{') {
+                            w.write(" ");
```

Comment:
are you not adding spaces to the raw literal here? Does this not actually change the value of the raw literal when parsed again?

Reply:
Yeah that is a bit hacky... It would produce unparsable bauble if we didn't though. So this state is not something that can come from parsed bauble.